### PR TITLE
Add aruco/3.1.12

### PR DESCRIPTION
**Pull request by bverhagen**
Specify library name and version:  **aruco/3.1.12**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/aruco/3.x.x/conandata.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+sources:
+  3.1.12:
+    url: https://downloads.sourceforge.net/project/aruco/3.1.12/aruco-3.1.12.zip?r=https%3A%2F%2Fsourceforge.net%2Fprojects%2Faruco%2Ffiles%2F3.1.12%2Faruco-3.1.12.zip%2Fdownload&ts=1604054853
```
Comment (madebr) [inline recipes/aruco/3.x.x/conandata.yml:3]:
```suggestion
    url: https://downloads.sourceforge.net/project/aruco/3.1.12/aruco-3.1.12.zip
```

File: recipes/aruco/3.x.x/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    @property
+    def _build_subfolder(self):
+        return "build_subfolder"
+
+    def requirements(self):
```
Comment (madebr) [inline recipes/aruco/3.x.x/conanfile.py:29]:
```suggestion
        return "build_subfolder"

    def config_options(self):
        if self.settings.os == "Windows":
            del self.options.fPIC

    def configure(self):
        if self.options.shared:
            del self.options.fPIC

    def requirements(self):
```

File: recipes/aruco/3.x.x/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        self.requires("opencv/4.5.0@")
+        self.requires("eigen/3.3.8@")
+        self.requires("zlib/1.2.11@")
```
Comment (madebr) [inline recipes/aruco/3.x.x/conanfile.py:32]:
```suggestion
        self.requires("opencv/4.5.0")
        self.requires("eigen/3.3.8")
        self.requires("zlib/1.2.11")
```

File: recipes/aruco/3.x.x/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            pattern="LICENSE", dst="licenses", src=self._source_subfolder
+        )
+
+        cmake = self._configure_cmake()
+        cmake.install()
+        tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
+        tools.rmdir(os.path.join(self.package_folder, "share"))
+
+    def package_info(self):
+        self.cpp_info.includedirs = ["include", "include/aruco"]
```
Comment (madebr) [inline recipes/aruco/3.x.x/conanfile.py:72]:
```suggestion
        self.cpp_info.includedirs.append(os.path.join("include", "aruco"))
```

File: recipes/aruco/3.x.x/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        cmake.build()
+
+    def package(self):
+        self.copy(
+            pattern="LICENSE", dst="licenses", src=self._source_subfolder
+        )
+
+        cmake = self._configure_cmake()
+        cmake.install()
+        tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
```
Comment (madebr) [inline recipes/aruco/3.x.x/conanfile.py:68]: Please also set the `pkg_config` generator name(s) in `package_info`.

File: recipes/aruco/3.x.x/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,7 @@
+cmake_minimum_required(VERSION 3.1)
+project(cmake_wrapper)
+
+include(${CMAKE_BINARY_DIR}/../conanbuildinfo.cmake)
```
Comment (madebr) [inline recipes/aruco/3.x.x/CMakeLists.txt:4]:
```suggestion
include(conanbuildinfo.cmake)
```

File: recipes/aruco/3.x.x/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        return "build_subfolder"
+
+    def requirements(self):
+        self.requires("opencv/4.5.0@")
+        self.requires("eigen/3.3.8@")
+        self.requires("zlib/1.2.11@")
+
+    def source(self):
+        tools.get(
+            **self.conan_data["sources"][self.version], filename="aruco.zip"
```
Comment (prince-chrismc) [inline recipes/aruco/3.x.x/conanfile.py:36]: usually we do not need filename, might be apart of madebr's suggestion for URL

File: recipes/aruco/3.x.x/test_package/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake"
+
+    settings = "os", "compiler", "arch", "build_type"
+    options = {"shared": [False, True], "fPIC": [False, True]}
+    default_options = {"shared": False, "fPIC": True}
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
```
Comment (prince-chrismc) [inline recipes/aruco/3.x.x/test_package/conanfile.py:15]: typically not required in the test_package

File: recipes/aruco/3.x.x/test_package/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+    def build(self):
```
Comment (madebr) [inline recipes/aruco/3.x.x/test_package/conanfile.py:17]: As @prince-chrismc  mentioned. `shared` and `fPIC` matter most when building libraries, not executables.
OTOH, I think `fPIC` is also useful for executables: `fPIE` for enaing ASLR.
```suggestion
    settings = "os", "compiler", "arch", "build_type"

    def build(self):
```

File: recipes/aruco/3.x.x/conanfile.py (hunk 7 above)
Comment (bverhagen) [reply recipes/aruco/3.x.x/conanfile.py:36]: After @madebr suggestions, this is indeed not necessary anymore.

File: recipes/aruco/3.x.x/conanfile.py (hunk 5 above)
Comment (bverhagen) [reply recipes/aruco/3.x.x/conanfile.py:74]: Done!

File: recipes/aruco/3.x.x/conanfile.py (hunk 4 above)
Comment (bverhagen) [reply recipes/aruco/3.x.x/conanfile.py]: I fixed this by appending both "include" and "include/aruco" in a similar way as your suggestion.

File: recipes/aruco/3.x.x/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        self._cmake.definitions["BUILD_GLSAMPLES"] = "OFF"
+        self._cmake.definitions["BUILD_UTILS"] = "OFF"
+        self._cmake.definitions["BUILD_DEBPACKAGE"] = "OFF"
+        self._cmake.definitions["BUILD_SVM"] = "OFF"
+        self._cmake.definitions["INSTALL_DOC"] = "OFF"
+        self._cmake.definitions["USE_OWN_EIGEN3"] = "OFF"
+        self._cmake.definitions["BUILD_SHARED_LIBS"] = self.options.shared
+        self._cmake.definitions[
+            "CMAKE_POSITION_INDEPENDENT_CODE"
+        ] = self.options.get_safe("fPIC", True)
```
Comment (prince-chrismc) [inline recipes/aruco/3.x.x/conanfile.py:61]: Question for the experts, is this the correct behavior? It's unusually so I would like confirmation

Comment (bverhagen) [reply recipes/aruco/3.x.x/conanfile.py:61]: I removed that again, as it is not fixing anything. I assume conan is already doing this under the hood.

Comment (madebr) [reply recipes/aruco/3.x.x/conanfile.py]: On Linux, when a shared library includes objects from a static library, that static library needs to be compiler with PIC on.I noticed this didn't happen by default, so I had to explicitly enable `CMAKE_POSITION_INDEPENDENT_CODE` when building a shared library.

File: recipes/aruco/3.x.x/test_package/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+
+set(CMAKE_CXX_STANDARD 11)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+find_package(aruco REQUIRED)
+
+add_executable(${PROJECT_NAME} test_package.cpp)
+target_link_libraries(${PROJECT_NAME} aruco::aruco)
```
Comment (madebr) [inline recipes/aruco/3.x.x/test_package/CMakeLists.txt:12]: Does aruco itself create a `Findaruca.cmake` file or `aruca-config.cmake` file?

**Review by madebr — APPROVED** [review]

File: recipes/aruco/3.x.x/test_package/CMakeLists.txt (hunk 11 above)
Comment (bverhagen) [reply recipes/aruco/3.x.x/test_package/CMakeLists.txt:12]: Yes, it provides an `arucoConfig.cmake` that provides a `aruco::aruco` target.

File: recipes/aruco/3.x.x/conanfile.py [hunk 12]
```diff
@@ -38,6 +38,12 @@ def requirements(self):
         self.requires("eigen/3.3.8")
         self.requires("zlib/1.2.11")
 
+        self.options[
+            "opencv"
+        ].shared = (
+            self.options.shared
+        )  # For some reason, this is not automagically inherited from self.options.shared
```
Comment (prince-chrismc) [inline recipes/aruco/3.x.x/conanfile.py:45]: does there need to be another check in build incase the user overrides? This is still very vague in my head, rather safe then sorry, if any one can confirm

**Review by prince-chrismc — APPROVED** [review]

File: recipes/aruco/3.x.x/conanfile.py (hunk 12 above)
Comment (madebr) [reply recipes/aruco/3.x.x/conanfile.py:45]: @bverhagen 
Will it fail if aruco is e.g. built as a static library and opencv as a shared library?
If so, then you should add a test to `build`, throwing a `ConanInvalidConfiguration` if you detect an error.

@uilianries 
Is there a conan way to do this available or on the horizon?

Comment (uilianries) [reply recipes/aruco/3.x.x/conanfile.py:45]: No, have to build and check.

> does there need to be another check in build incase the user overrides

No, even if it was in configure(), all user options are collected and available in configure.

Comment (uilianries) [reply recipes/aruco/3.x.x/conanfile.py:45]: I'm curious about your error, it seems be related to link order

Comment (bverhagen) [reply recipes/aruco/3.x.x/conanfile.py:45]: This is the error:

```/usr/bin/ld: bin/test_package: hidden symbol `_Z7cvTruncRKN2cv9softfloatE' in /home/conan/w/cci_PR-3646/.conan/data/opencv/4.5.0/_/_/package/74bf96ed8c4193fbaf98035f279434691d003ce6/lib/libopencv_core.a(softfloat.cpp.o) is referenced by DSO
/usr/bin/ld: final link failed: Bad value    
collect2: error: ld returned 1 exit status    
make[2]: *** [bin/test_package] Error 1    
make[1]: *** [CMakeFiles/test_package.dir/all] Error 2```

I don't think it is related to link order, but I am prepared to be proven wrong :-)

Comment (prince-chrismc) [reply recipes/aruco/3.x.x/conanfile.py:45]: Hmmm https://stackoverflow.com/a/23708713/8480874 never seen "hidden linkage" something new to learn!

Comment (madebr) [reply recipes/aruco/3.x.x/conanfile.py:45]: Is the hiding of the symbol a big in opencv (not exporting the symbol) or is this a case of using a hidden symbol (and depending on an implementation detail)?

**Review by SSE4 — APPROVED** [review]

**Pull request merged**